File: 217-contains-duplicate.py
# https://leetcode.com/problems/contains-duplicate/description/
# Do not use a list in this case, lookups are faster in dictionaries!
class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        val_times_map = {} #Use dict for lookup!
        for num in nums:
            if num in val_times_map.keys():
                return True
            val_times_map[num] = 1
        return False

# Lookups are faster in dictionaries because Python implements them using hash tables. 
# If we explain the difference by Big O concepts, 
# dictionaries have constant time complexity, O(1) while lists have linear time complexity, O(n).

File: cpp/217-contains-duplicate.cpp
class Solution {
public:
    bool containsDuplicate(vector<int>& nums) {
        // Using unordered_set for the find time: complexity O(1)
        std::unordered_set<int> check;
        for (const auto&n: nums){
            if (check.find(n) != check.end()){ // the number is in the set
                return true;
            }
            else{
                check.insert(n);
            }
        }
        return false;
    }
};